src/core/ad.jl

```diff
@@ -1,26 +1,96 @@
+abstract type ADBackend end
+struct ForwardDiffAD{chunk} <: ADBackend end
+struct FluxTrackerAD <: ADBackend end
+struct NoAD <: ADBackend end
+ADBackend() = ADBackend(ADBACKEND[])
+ADBackend(T::Symbol) = ADBackend(Val(T))
+function ADBackend(::Val{T}) where {T}
+    if T === :forward_diff
+        return ForwardDiffAD{CHUNKSIZE[]}
+    else
+        return FluxTrackerAD
+    end
+end
+
+getchunksize(::Type{ForwardDiffAD{chunk}}) where chunk = Val(chunk)
+
+
+"""
+getADtype(alg)
+
+Finds the autodifferentiation type of the algorithm `alg`.
+"""
+function getADtype()
+    if ADBACKEND[] == :forward_diff
+        return ForwardDiffAD{CHUNKSIZE[]}
+    else
+        return FluxTrackerAD
+    end
+end
+getADtype(::Any) = NoAD
+getADtype(s::Sampler) = getADtype(typeof(s))
+getADtype(s::Type{<:Sampler{TAlg}}) where {TAlg} = getADtype(TAlg)
+getADtype(alg::Union{Hamiltonian, Gibbs, PMMH}) = getADtype(typeof(alg))
+getADtype(::Type{<:Hamiltonian{AD}}) where {AD} = AD
+
+@generated function getADtype(::Union{Type{<:Gibbs{A}}, Type{<:PMMH{T, A}}}) where {T, A <: Tuple}
```

Comment:
Can we add some comments to say what's the logic of this function. E.g. if no sampling alg uses AD, returns NoAD; if only one sampling alg uses AD, returns its type and if more than one uses AD, return only if they are consistent otherwise throw an error.

Reply:
Also we can probably support mixed AD if we tie the AD type to samplers?